ic-agent/src/agent/agent_test.rs

```diff
@@ -236,3 +237,138 @@ fn status_error() -> Result<(), AgentError> {
+    221, 155, 201, 119, 224, 93, 99, 40, 248, 137, 220, 78, 124, 153, 17, 76, 115, 122, 73, 70, 83,
+    203, 39, 161, 245, 92, 6, 244, 85, 94, 15, 22, 9, 128, 175, 94, 173, 9, 138, 204, 25, 80, 16,
+    178, 247,
+];
+
+#[test]
+// asserts that a delegated certificate with correct /subnet/<subnetid>/canister_ranges
+// passes the certificate verification
+fn check_subnet_range_with_valid_range() -> Result<(), AgentError> {
+    let _read_mock = mock(
+        "POST",
+        "/api/v2/canister/ivg37-qiaaa-aaaab-aaaga-cai/read_state",
+    )
+    .with_status(200)
+    .with_body(&REQ_WITH_DELEGATED_CERT_RESPONSE)
+    .create();
+    let agent = Agent::builder()
+        .with_transport(ReqwestHttpReplicaV2Transport::create(
+            &mockito::server_url(),
+        )?)
+        .build()?;
+    let runtime = tokio::runtime::Runtime::new().expect("Unable to create a runtime");
+    let result = runtime.block_on(async {
+        agent
+            .read_state_raw(
+                vec![REQ_WITH_DELEGATED_CERT_PATH
+                    .iter()
+                    .map(Label::from)
+                    .collect()],
+                Principal::from_text(REQ_WITH_DELEGATED_CERT_CANISTER).unwrap(),
+            )
+            .await
+    });
+    match result {
+        Ok(_) => Ok(()),
+        Err(other) => panic!("found something unexpected: {:?}", other),
+    }
+}
+
+#[test]
+// asserts that a delegated certificate with /subnet/<subnetid>/canister_ranges that don't include
+// the canister gets rejected by the cert verification because the subnet is not authorized to
+// respond to requests for this canister. We do this by using a correct response but serving it
+// for the wrong canister, which a malicious node might do.
+fn check_subnet_range_with_unauthorized_range() -> Result<(), AgentError> {
+    let wrong_canister = Principal::from_text("ryjl3-tyaaa-aaaaa-aaaba-cai").unwrap();
+    let _read_mock = mock(
+        "POST",
+        "/api/v2/canister/ryjl3-tyaaa-aaaaa-aaaba-cai/read_state",
+    )
+    .with_status(200)
+    .with_body(&REQ_WITH_DELEGATED_CERT_RESPONSE)
+    .create();
+    let agent = Agent::builder()
+        .with_transport(ReqwestHttpReplicaV2Transport::create(
+            &mockito::server_url(),
+        )?)
+        .build()?;
+    let runtime = tokio::runtime::Runtime::new().expect("Unable to create a runtime");
+    let result = runtime.block_on(async {
+        agent
+            .read_state_raw(
+                vec![REQ_WITH_DELEGATED_CERT_PATH
+                    .iter()
+                    .map(Label::from)
+                    .collect()],
+                wrong_canister,
+            )
+            .await
+    });
+    match result {
+        Err(AgentError::CertificateNotAuthorized()) => Ok(()),
+        Ok(_) => panic!("cert with bad canister_range considered valid"),
+        Err(other) => panic!("found something unexpected: {:?}", other),
+    }
```

Comment:
nit: we could simplify this a bit:
```suggestion
    assert_eq!(result, Err(AgentError::CertificateNotAuthorized()));
```
I believe the error message in case of a test failure will be clear enough.